pkg/dependencies/factory.go

```diff
@@ -0,0 +1,391 @@
+package dependencies
+
+import (
+	"context"
+	"errors"
+
+	"github.com/aws/eks-anywhere/pkg/addonmanager/addonclients"
+	"github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/bootstrapper"
+	"github.com/aws/eks-anywhere/pkg/clients/flux"
+	"github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/clustermanager"
+	"github.com/aws/eks-anywhere/pkg/executables"
+	"github.com/aws/eks-anywhere/pkg/filewriter"
+	"github.com/aws/eks-anywhere/pkg/networking"
+	"github.com/aws/eks-anywhere/pkg/providers"
+	"github.com/aws/eks-anywhere/pkg/providers/factory"
+	supportbundle "github.com/aws/eks-anywhere/pkg/support"
+)
+
+type Dependencies struct {
+	Provider         providers.Provider
+	ClusterAwsCli    *executables.Clusterawsadm
+	DockerClient     *executables.Docker
+	Kubectl          *executables.Kubectl
+	Govc             *executables.Govc
+	Writer           filewriter.FileWriter
+	Kind             *executables.Kind
+	Clusterctl       *executables.Clusterctl
+	Flux             *executables.Flux
+	Troubleshoot     *executables.Troubleshoot
+	Networking       clustermanager.Networking
+	ClusterManager   *clustermanager.ClusterManager
+	Bootstrapper     *bootstrapper.Bootstrapper
+	FluxAddonClient  *addonclients.FluxAddonClient
+	AnalyzerFactory  supportbundle.AnalyzerFactory
+	CollectorFactory supportbundle.CollectorFactory
+}
+
+func ForSpec(ctx context.Context, clusterSpec *cluster.Spec) *Factory {
+	eksaToolsImage := clusterSpec.VersionsBundle.Eksa.CliTools
+	return NewFactory().
+		WithExecutableBuilder(ctx, clusterSpec.UseImageMirror(eksaToolsImage.VersionedImage())).
+		WithWriterFolder(clusterSpec.Name).
+		WithDiagnosticCollectorImage(clusterSpec.VersionsBundle.Eksa.DiagnosticCollector.VersionedImage())
```

Comment:
Yeah, this method is just a helper when creating dependencies based on a particular `clusterSpec`. In that case, there are certain configurations that are always the same. The collector image is not always used but is always the same for the same `clusterSpec`, that's why is here.
To be precise, these `With*` methods in the helper are slightly different than the other ones, since they don't necessarily build dependencies, but they configure the factory to be able to build the real dependencies.

Regarding `New()`, I believe that's`NewFactory()`, isn't it? (it' called `NewFactory` instead of just `New` because it doesn't build the `dependencies` but a factory to build dependencies)

Reply:
I get the distinction now, this looks good to me